### ISSUE TITLE
unified sound registry

### ISSUE THREAD

**Issue by felixroos** [tidalcycles/strudel#233]
This might not be a super concrete issue (yet), but I feel there is something missing in how sounds are handled

currently, all available sounds are..

1. one of the default waveforms `sine` `square` `triangle` `sawtooth`
2. one of the hard coded soundfonts (see [this list](https://github.com/tidalcycles/strudel/blob/main/packages/soundfonts/list.mjs#L1))
3. one of the default samples loaded in [prebake](https://github.com/tidalcycles/strudel/blob/main/repl/src/prebake.mjs). `piano` or one of [this list](https://strudel.tidalcycles.org/EmuSP12.json)
4. one of the samples loaded by the user via `samples(...)` ([custom sample maps](https://strudel.tidalcycles.org/learn/samples/))

ideas of what could be possible in the future:

- a) sf2 files populating the sound registry (currently a separate param see https://strudel.tidalcycles.org/?DHUbrEloJxMd ) https://github.com/tidalcycles/strudel/issues/182
- b) a way to define "presets" as a mapping of a sound name to a pattern . An alternative to how the `.piano()` preset is defined [here](https://github.com/tidalcycles/strudel/blob/main/repl/src/prebake.mjs#L54) https://github.com/tidalcycles/strudel/issues/462

b is similar to what @yaxu suggested in the discord a while ago:

"It could work like this:

```js
let mysounds = inhabit({
"bd": speed("1 1.5 2 1").s('https://samplepack.com/bd.wav').begin(0).end(0.5).squiz(1.1), 
"sd": s('https://samplepack.com/sd.wav').squiz(1.5)
}) 
mysounds("bd sd")
```

In general I like the idea of keeping everything in the pattern where possible, rather than adding external datastructures/metadata that can't then be patterned freely".

We could also generalize `samples` to `sounds` like this:

```js
sounds({
  bd: x => x.sample('bd/BT0AADA.wav').squiz(1.1),
  fatbass: x => x.add(note(0.05)).s('sawtooth') // detuned saw
}, );
sampleRoot('https://raw.githubusercontent.com/tidalcycles/Dirt-Samples/master/') // not sure about that line

stack(
  note("c3 eb3 g3").s('fatbass'),
  s("bd*4")
)
```

above would be equivalent to

```js
stack(
  note("c3 eb3 g3").add(note(0.05)).s('sawtooth'),
  s("bd*4").sample('bd/BT0AADA.wav').squiz(1.1)
)
```

Just some ideas..

**Comment by felixroos** [comment]
@yaxu idea from chat: be able to load samples from json url `samples('https://mywebsite.org/soundbanks.json')`. possible format:

```js
  const machines = [
     {
      bank: "AJKPercusyn",
      base:
        "https://raw.githubusercontent.com/ritchse/tidal-drum-machines/main/machines/AJKPercusyn/",
      license: "CC-BY (or whatever it is) (c) J. Doe",
      samples: {
      bd: ["ajkpercusyn-bd/Bassdrum.wav"],
      cb: ["ajkpercusyn-cb/Cowbell.wav", "ajkpercusyn-cb/Snarepop.wav"],
      ht: ["ajkpercusyn-ht/Tom.wav"],
      sd: ["ajkpercusyn-sd/Noise.wav"],
      break: [{sample: "amen/amen.wav", cuepoints: [0.2, 0.52]}]
      }
    },
```

I wonder if we should generally store loaded samples like that or use it just as one possible input format (keeping the existing in memory representation). Having more metadata might be useful when listing loaded samples in the front end

**Comment by felixroos** [comment]
just a quick idea on banks: if no `.bank` is specified, the pool of samples could be all samples with that name in all banks combined. example:

```js
[
  { bank: '808', samples: { bd: ['808/bd.wav'] },
  { bank: '909', samples: { bd: ['909/bd.wav'] }
]
```

```js
s("bd") // uses samples ['808/bd.wav', '909/bd.wav' ]
s("bd").bank('808') // uses samples ['808/bd.wav']
s("bd").bank('909') // uses samples ['909/bd.wav']
```

so the bank could act as a filter to all registered samples..

**Comment by yaxu** [comment]
Yes that makes me think that if a 'bank' is a way to tag samples to filter them on, maybe you could tag samples in other ways that you could make up on the spot and then filter on later. Might well be taking it too far though.

**Comment by felixroos** [comment]
this issue overlaps with https://github.com/tidalcycles/strudel/issues/182 and https://github.com/tidalcycles/strudel/issues/462

**Comment by felixroos** [comment]
this issue is a bit convoluted / hard to parse... the original idea was solved in https://github.com/tidalcycles/strudel/pull/516

the rest now lives in

- https://github.com/tidalcycles/strudel/issues/182
- https://github.com/tidalcycles/strudel/issues/462
- https://github.com/tidalcycles/strudel/issues/540
